File: 2020/day/5/main.py
import fileinput

def parse_seat(s):
    row_lo, row_hi = 0, 127
    seat_lo, seat_hi = 0, 7

    for x in s:
        if x == 'B':
            row_lo = (row_lo + row_hi + 1) // 2
        elif x == 'F':
            row_hi = row_lo + (row_hi - row_lo) // 2
        elif x == 'R':
            seat_lo = (seat_lo + seat_hi + 1) // 2
        elif x == 'L':
            seat_hi = seat_lo + (seat_hi - seat_lo) // 2
            
    assert row_lo == row_hi, (s, row_lo, row_hi)
    assert seat_lo == seat_hi, (s, seat_lo, seat_hi)
    return row_lo, seat_lo

def seat_id(seat):
    row, col = seat
    return row * 8 + col

seat_ids = set(seat_id(parse_seat(line.strip())) for line in fileinput.input('input.txt'))

# Part 1
print(max(seat_ids))

# Part 2
for i in range(min(seat_ids), max(seat_ids)):
    if i not in seat_ids and i - 1 in seat_ids and i + 1 in seat_ids:
        print(i)

File: 2020/day/5/Main.hs
import Data.List
import qualified Data.IntSet as IntSet

data Range a = Range !a !a
    deriving (Eq, Ord, Show)

upperHalf (Range lo hi) = Range ((lo + hi + 1) `div` 2) hi
lowerHalf (Range lo hi) = Range lo (lo + (hi - lo) `div` 2)

data Seat a = Seat (Range a) (Range a)
    deriving (Eq, Ord, Show)

step (Seat r c) 'B' = Seat (upperHalf r) c
step (Seat r c) 'F' = Seat (lowerHalf r) c
step (Seat r c) 'R' = Seat r (upperHalf c)
step (Seat r c) 'L' = Seat r (lowerHalf c)

scan :: String -> Seat Int
scan = foldl step (Seat (Range 0 127) (Range 0 7))

seatId :: Seat Int -> Int
seatId (Seat (Range r _) (Range c _)) = r * 8 + c

main = do
    input <- readFile "input.txt"
    let seats = map scan (lines input)
        seatIds = IntSet.fromList (map seatId seats)

    -- Part 1
    let Just (maxSeatId, _) = IntSet.maxView seatIds
    print maxSeatId

    -- Part 2
    let Just (minSeatId, _) = IntSet.minView seatIds
        Just yourSeatId = find (`IntSet.notMember` seatIds) [minSeatId..maxSeatId]
    print yourSeatId